### PR TITLE
Plugins

### PR DESCRIPTION
**Pull request by delvedor**
Finally, the very first implementation of the plugin system.
Behind the scenes me and @mcollina worked hard to achieve the following features:
- Plugin encapsulation
- Plugin Inheritance
- Asynchronous plugin load
- Dependency management

There is still lot of work to do, such as find nicer names for the apis, think about a better system for dependency handling (name and version?) and if we want to expose some internals to the plugins.

We also need to figure out how can we add methods to `Request` and `Reply` as well (we not necessarily have to do it in this PR).

Example:
```js
const fastify = require('fastify')()

// register provides encapsulation
fastify.register((instance, opts, next) => {
  // plugin do the same work of register, but without encapsulation
  instance.plugin((i, o, n) => {
    i.addServerMethod('test', 'first')
    n()
  })

  instance.get('/first', (req, reply) => {
    // this will always send { plugin: 'first' }
    reply.send({ plugin: instance.test })
  })

  next()
})

fastify.register((instance, opts, next) => {
  // even if they have the same name, the two plugins does not overwrite each other
  instance.plugin((i, o, n) => {
    i.addServerMethod('test', 'second')
    n()
  })

  instance.get('/second', (req, reply) => {
    // this will always send { plugin: 'second' }
    reply.send({ plugin: instance.test })
  })

  next()
})

fastify.listen(3000, err => {
  if (err) throw err
})
```
cc @davidmarkclements @lucamaraschi @jsumners

### CODE REVIEW

**Comment by davidmarkclements** [comment]
nice!

if they have the same name, but don't overwrite each other - what's the purpose of the name?

If logging only, then maybe we should warn - since that would create confusing logs? 

what do `i` `o` and `n` (guessing `n` is next/cb is `i` instance?) represent?

why would `first` be passed to `addServerMethod` - shouldn't the second param be a function? If not I'm confused as to what addServerMethod does? 

instead of `addServerMethod(name, val)` what if we have `add(type, name, val)` where type can be `server` (and other things)  - instead of `add` maybe `addTo`, or `enhance` or `decorate`

**Comment by delvedor** [comment]
> if they have the same name, but don't overwrite each other - what's the purpose of the name?

I don't think I've understand the question.

> If logging only, then maybe we should warn - since that would create confusing logs?

Same as above.

> what do `i` `o` and `n` (guessing `n` is next/cb is `i` instance?) represent?

Yeah, they are just the fastify instance, the options and the next callback.
Why the different naming? Good question :P

> why would `first` be passed to `addServerMethod` - shouldn't the second param be a function? If not I'm confused as to what addServerMethod does?

Yeah, it is intended to be a function, but for testing I used a string, once the api is finalized we should add a type check unless we want this behaviour (maybe the user wants to save something in the server instance and access it later)

> instead of `addServerMethod(name, val)` what if we have `add(type, name, val)` where type can be `server` (and other things) - instead of `add` maybe `addTo`, or `enhance` or `decorate`

This was my original idea, but it requires a lot of work handle the update of request and reply, so we decided to start with the server instance and then support request/reply as well.
👍  for `decorate`!

**Comment by jsumners** [comment]
Do I understand this correctly?: `fastify.register` registers "middleware" functions within which all `fastify.plugin` invocations are scoped to a copy of the `fastify` instance. Whereas `fastify.plugin` registers "middleware" on whichever instance it is invoked.

Thus:

```js
test('fastify.plugin registers root level plugins', t => {
  t.plan(15)
  const fastify = Fastify()

  fastify.plugin((instance, opts, next) => {
    instance.addServerMethod('test', 'first')
    t.ok(instance.test)
    next()
  })

  fastify.register((instance, opts, next) => {
    t.ok(instance.test)
    instance.plugin((i, o, n) => {
      i.addServerMethod('test2', 'second')

      i.get('/test2', (req, reply) => {
        t.ok(i.test2)
        reply.send({ test2: i.test2 })
      })

      n()
    })
    next()
  })

  fastify.ready(() => {
    t.ok(fastify.test)
    t.notOk(fastify.test2)
  })

  fastify.get('/', (req, reply) => {
    t.ok(fastify.test)
    reply.send({ test: fastify.test })
  })

  fastify.listen(0, err => {
    t.error(err)
    fastify.server.unref()

    request({
      method: 'GET',
      uri: 'http://localhost:' + fastify.server.address().port
    }, (err, response, body) => {
      t.error(err)
      t.strictEqual(response.statusCode, 200)
      t.strictEqual(response.headers['content-length'], '' + body.length)
      t.deepEqual(JSON.parse(body), { test: 'first' })
    })

    request({
      method: 'GET',
      uri: 'http://localhost:' + fastify.server.address().port + '/test2'
    }, (err, response, body) => {
      t.error(err)
      t.strictEqual(response.statusCode, 200)
      t.strictEqual(response.headers['content-length'], '' + body.length)
      t.deepEqual(JSON.parse(body), { test2: 'second' })
    })
  })
})
```

**Comment by delvedor** [comment]
@jsumners everything you do inside a `.register` is "scoped", in other word, the `fastify` instance passed by register is always overridden with `Object.create(fastify)`.
`.plugin` work in the exact same way as register, but the `fastify` instance is not overridden.

This because we must use the [avvio](https://github.com/mcollina/avvio) lifecycle to ensure that everything is loaded before the server boot up.

Your example is correct, you get the point.
But the`.get` inside the `.plugin` is not 100% correct, it should be outside, check [here](https://github.com/fastify/fastify/blob/2d084f6b4d3c0418a43a5bf42fa33b0ea9d7ff2b/test/plugin.test.js#L18-L33).

**Comment by delvedor** [comment]
Hi guys, as you can see we did some changes with the api.
Now for load either plugins, routes or whatever we use only `.register`.
By default `register` creates an *encapsulation*, so if you are adding methods to the server (via the new *decorator* api), those methods will be available only inside that `register` scope (and its sons).

We did some changes to `avvio`, the system that we use to load asynchronously everything before start the server. Every time we use `.register`, internally avvio creates a new instance of Fastify with `Object.create` (check the override function [here](https://github.com/fastify/fastify/blob/0310694e97884a03c29818a38d3bd83ac8112906/fastify.js#L46-L52)).

@mcollina and me thought that have two different api for register stuff to Fastify is confusing and we agreed that we need only one api to do this, `register`.
But, as I wrote few line above, register creates by default a new scope every time, so all the plugins loaded with a register will not be available to the correct scope.
example:
```js
module.exports = function (fastify, opts, next) {
  fastify.register(function (fastify, opts, next) {
    fastify.decorate('test', () => {})
    next()
  })

  fastify.get('/', (req, reply) => {
    // here fastify.test will not be available
    reply.send(...)
  })

  next()
}
```
To solve this the plugin developer has two ways:
- use the hidden property `'skip-override'` with `Symbol`
- use the `fastify-plugin` utility (check [here](https://github.com/fastify/fastify-plugin) to see how it works).

Example:
```js
module.exports = function (fastify, opts, next) {
  fastify.register(require('some-plugin.js'))

  fastify.get('/', (req, reply) => {
    // here fastify.test will be available
    reply.send(...)
  })

  next()
}
```
```js
const fp = require('fastify-plugin')
module.exports = fp(function (fastify, opts, next) {
  fastify.decorate('test', () => {})
  next()
})
```
In this way we can meet our goals:
- Plugin encapsulation
- Plugin Inheritance
- Asynchronous plugin load
- Dependency management

And use Avvio to ensure that everything is correctly loaded before the server starts (hint, the server start once avvio emits the `'ready'` event).

I hope I explained myself, if not, please let me know.
If you have time, feedback are appreciated.

**Comment by mcollina** [comment]
This is a massive one. Would you mind adding docs as well? :)

**Comment by delvedor** [comment]
Shall we add a strict function check for the fastify decorator?
Without the check a user can save everything he needs in a fastify property, such as configurations or whatever.
Thoughts?
____
> This is a massive one. Would you mind adding docs as well? :)

@mcollina  docs will arrive soon :)

**Comment by mcollina** [comment]
@delvedor I don't think we should just support decorating with functions, let's leave it open for now.

**Comment by delvedor** [comment]
Here the docs!
Feedback appreciated :)

File: docs/Decorators.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<h1 align="center">Fastify</h1>
+
+## Decorators
+
+If you need to add functionalities to the Fastify instance, the `decorate` api is what you need.
+
+This api allows you to add new properties to the Fastify instance, a property value is not restricted to be a function, could also be an object or a string for example.
+
+*Note that `decorate` will add new properties to the Fastify instance, not Reply or Request objects.*
```
Comment (mcollina) [inline docs/Decorators.md]: Maybe we can say, _future versions of Fastify will support that_.

File: docs/Plugins.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+We recommend to use the `fastify-plugin` module.
+```js
+const fp = require('fastify-plugin')
+
+module.exports = fp(function (fastify, opts, next) {
+  // your plugin code
+  next()
+})
+```
```
Comment (mcollina) [inline docs/Plugins.md]: this information is repeated in the next section, maybe we can just link down to the relative sections in the list.

File: docs/Plugins.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  if (err) throw err
+})
+```
+
+By default, `register` creates a *new scope*, this means that everything you do to the current Fastify instance is not reflected to the current context ancestors, but only to its sons. This feature allows us to achieve plugin *encapsulation* and *inheritance*.
+
+
+### Use a plugin
+Use a plugin is very easy, you just need to require it and pass it as parameter to the `register` API.
+You can pass a single plugin or an array of plugins if all of them can share the same options object/callback.
```
Comment (mcollina) [inline docs/Plugins.md]: Using a plugin is covered in the above.

File: docs/Plugins.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Sometimes you don't want to use the plugin *encapsulation*, in that cases you have two ways to tell Fastify to skip the creation of a new scope:
```
Comment (mcollina) [inline docs/Plugins.md]: this does not explain why a plugin author is responsible for controlling the scope of how his plugin is used.

File: fastify.js [hunk 5]
```diff
@@ -41,7 +42,14 @@ function build (options) {
...
 
-  pluginLoader(fastify, {})
+  const app = avvio(fastify, {})
+  // Override to allow the plugin incapsulation
+  app.override = function (server, fn) {
+    if (fn[Symbol.for('skip-override')]) {
+      return server
+    }
+    return Object.create(server)
+  }
```
Comment (mcollina) [inline fastify.js]: can we move this to a top-level function downwards?

File: docs/Plugins.md (hunk 3 above)
Comment (delvedor) [inline docs/Plugins.md]: Maybe we can completely remove the section "Use a plugin", move part of its content in the section above and add to the "Create a plugin" section another example without `fastify-plugin`.

Comment (mcollina) [inline docs/Plugins.md]: We need a "use a plugin" section. What should a user expect from a plugin?

Comment (delvedor) [inline docs/Plugins.md]: > Fastify allows the user to extend its functionalities with plugins. A plugin can be a set of routes, a server decorator or whatever. The API that you will need to use one or more plugins, is register.

Do you think that we must expand this paragraph?

Comment (mcollina) [inline docs/Plugins.md]: A little bit, yes. We should also reference decorate, etc.. maybe linking there is enough.
We should explain the principles behind encapsulation here (it's ok repeating those)

**Comment by mcollina** [comment]
We need to have an hook to support a safe shutdown of the plugins, we need this to close connections, etc in a proper way.

**Comment by delvedor** [comment]
@mcollina with the latest commit we support the `'onClose'` hook.
I'm not that sure if we must listen for the `'close'` event of the server as well.

**Comment by mcollina** [comment]
That's perfect but docs :).

**Pull request merged**